crates/compiler/builtins/roc/Json.roc

```diff
@@ -463,3 +476,82 @@ decodeRecord = \initialState, stepField, finalizer -> Decode.custom \bytes, @Jso
         when finalizer endStateResult is
             Ok val -> { result: Ok val, rest: afterRecordBytes }
             Err e -> { result: Err e, rest: afterRecordBytes }
+
+# Helper to eat leading Json whitespace characters
+eatWhitespace = \input ->
+    when input is
+        [' ', ..] -> eatWhitespace (List.dropFirst input)
+        ['\n', ..] -> eatWhitespace (List.dropFirst input)
+        ['\r', ..] -> eatWhitespace (List.dropFirst input)
+        ['\t', ..] -> eatWhitespace (List.dropFirst input)
+        _ -> input
+
+# Test eating Json whitespace
+expect
+    input = Str.toUtf8 " \n\r\tabc"
+    actual = eatWhitespace input
+    expected = Str.toUtf8 "abc"
+    actual == expected
+
+# Test json string decoding with escapes
+expect
+    input = Str.toUtf8 "\"a\r\nbc\\\"xz\""
+    expected = Ok "a\r\nbc\\\"xz"
+    actual = Decode.fromBytes input fromUtf8
+    actual == expected
+
+# Test json string encoding with escapes
+expect
+    input = "a\r\nbc\\\"xz"
+    expected = Str.toUtf8 "\"a\r\nbc\\\"xz\""
+    actual = Encode.toBytes input toUtf8
+    actual == expected
+
+# Test json array decode empty list
+expect
+    input = Str.toUtf8 "[ ]"
+    expected = []
+
+    actual : List U8
+    actual = Decode.fromBytes input fromUtf8 |> Result.withDefault []
+
+    actual == expected
+
+# Test json array decoding into integers
+expect
+    input = Str.toUtf8 "[ 1,\n2,\t3]"
+    expected = [1, 2, 3]
+
+    actual : List U8
+    actual = Decode.fromBytes input fromUtf8 |> Result.withDefault []
+
+    actual == expected
```

Comment:
🎨 No need to change, but as a general rule I'd try to preserve the `Result` like so:

```elm
    expected = Ok [1, 2, 3]

    actual : Result (List U8) _
    actual = Decode.fromBytes input fromUtf8

    actual == expected
```

This way if it fails because the operation returned an `Err`, we can see that it was an `Err` (along with any diagnostic info the `Err` might contain) when `actual` gets printed out.